### PR TITLE
Add refined inteface to rating graph and corresponding tests

### PR DESCRIPTION
**Pull request by MkuuWaUjinga**

### CODE REVIEW

File: util/datastructures.py [hunk 1]
```diff
@@ -6,22 +10,89 @@ def __init__(self):
         self.edges = []
         self.nodes = []
 
-    def __init__(self, nodes, edges):
-        assert (len(nodes) - 1 == len(edges)) or (len(nodes) == 0 and len(edges) == 0), "Path is not valid, number of edges and nodes does not compse a path"
+    def __init__(self, nodes:List, edges:List):
```
Comment (Baschdl) [inline util/datastructures.py]: Can we define the type of the nodes and edges in the lists here?

File: util/datastructures.py [hunk 2]
```diff
@@ -6,22 +10,89 @@ def __init__(self):
...
-    def as_list(self):
+    def as_list(self) -> List:
         representation = [None] * len(self)
         representation[::2] = self.nodes
         representation[1::2] = self.edges
         return representation
 
-    def __len__(self):
+    @staticmethod
+    def from_list(self, meta_path: List):
```
Comment (Baschdl) [inline util/datastructures.py]: Typing at least for the return value possible.

File: util/datastructures.py [hunk 3]
```diff
@@ -6,22 +10,89 @@ def __init__(self):
...
+        assert len(meta_path) % 2 != 0 or len(meta_path) == 0
+        return MetaPath(meta_path[::2], meta_path[1::2])
+
+    @staticmethod
+    def from_string(self, meta_path: str, sep=' '):
```
Comment (Baschdl) [inline util/datastructures.py]: You can even type variables with standard values 😁

File: util/datastructures.py [hunk 4]
```diff
@@ -6,22 +10,89 @@ def __init__(self):
...
+    def __init__(self):
+        self.g = Graph(directed=True)
+        self.distance = self.g.new_edge_property("double")
+        self.meta_paths_map = {}
+
+    """
+    :param a: The index of the meta-path will be retrieved or the meta-path is mapped to a new index.
+    """
+
+    def __add_meta_path(self, a: MetaPath):
```
Comment (Baschdl) [inline util/datastructures.py]: Variable name `a` could be a bit more descriptive.

File: util/datastructures.py [hunk 5]
```diff
@@ -6,22 +10,89 @@ def __init__(self):
...
+    domain_value = None
+
+    def __init__(self, meta_path: MetaPath):
+        self.meta_path = meta_path
+
+
+class MetaPathRatingGraph:
+    meta_paths_map = None
+    distance = None
+    g = None
```
Comment (GittiHab) [inline util/datastructures.py]: That's not very descriptive. Can we change this to a more descriptive name? For example `graph`

File: util/datastructures.py [hunk 6]
```diff
@@ -6,22 +10,89 @@ def __init__(self):
...
+        self.distance = self.g.new_edge_property("double")
+        self.meta_paths_map = {}
+
+    """
+    :param a: The index of the meta-path will be retrieved or the meta-path is mapped to a new index.
+    """
+
+    def __add_meta_path(self, a: MetaPath):
+        if a in self.meta_paths_map:
+            v = self.meta_paths_map[a]
```
Comment (GittiHab) [inline util/datastructures.py]: `v` could also be more descriptive. I have no idea what this is.

File: util/datastructures.py [hunk 7]
```diff
@@ -6,22 +10,89 @@ def __init__(self):
...
+    meta_path = None
+    structural_value = None
+    domain_value = None
+
+    def __init__(self, meta_path: MetaPath):
+        self.meta_path = meta_path
+
+
+class MetaPathRatingGraph:
+    meta_paths_map = None
```
Comment (GittiHab) [inline util/datastructures.py]: What does this map to what? Meta-paths to `?`...

File: util/datastructures.py [hunk 8]
```diff
@@ -6,22 +10,89 @@ def __init__(self):
...
+
+    def __add_meta_path(self, a: MetaPath):
+        if a in self.meta_paths_map:
+            v = self.meta_paths_map[a]
+        else:
+            v = self.g.add_vertex()
+            self.meta_paths_map[a] = v
+        return v
+
+    """
```
Comment (GittiHab) [inline util/datastructures.py]: We usually put the comments right after the method definition, so

    add_user_rating(...):
    """
    comment
    """

File: util/datastructures.py [hunk 9]
```diff
@@ -6,22 +10,89 @@ def __init__(self):
...
+    :param a: The meta-path, which was rated higher compared to b.
+    :param b: The meta-path, which was rated lower compared to a.
+    :param distance: The distance between meta-paths a and b. 
+    """
+
+    def add_user_rating(self, a: MetaPath, b: MetaPath, distance: float):
+        id_a = self.__add_meta_path(a)
+        id_b = self.__add_meta_path(b)
+
+        new_edge_positive = self.g.add_edge(id_a, id_b)
```
Comment (GittiHab) [inline util/datastructures.py]: Maybe add an assertion before this to make sure that `distance` is positive. If it is negative swap a and b and then multiply `distance` with `-1`.

File: util/datastructures.py [hunk 10]
```diff
@@ -6,22 +10,89 @@ def __init__(self):
...
+            self.meta_paths_map[a] = v
+        return v
+
+    """
+    :param a: The meta-path, which was rated higher compared to b.
+    :param b: The meta-path, which was rated lower compared to a.
+    :param distance: The distance between meta-paths a and b. 
+    """
+
+    def add_user_rating(self, a: MetaPath, b: MetaPath, distance: float):
```
Comment (GittiHab) [inline util/datastructures.py]: ...maybe change `a` and `b` to `metapath_a` and `metapath_b`...

File: util/datastructures.py [hunk 11]
```diff
@@ -6,22 +10,89 @@ def __init__(self):
...
+        return self.g.num_edges()
+
+    def number_of_nodes(self) -> int:
+        return self.g.num_vertices()
+
+    def all_nodes(self) -> List[MetaPath]:
+        return list(self.meta_paths_map.keys())
+
+    def all_pair_distances(self):
+        return shortest_distance(self.g, weights=self.distance, directed=True)
```
Comment (GittiHab) [inline util/datastructures.py]: Where is `shortest_distance` defined?

File: Tests/MetaPathRatingGraphTest.py [hunk 12]
```diff
@@ -0,0 +1,69 @@
+import unittest
+from typing import List
+from util.datastructures import MetaPathRatingGraph, MetaPath
+
+
+class MetaPathRatingTest(unittest.TestCase):
+
+    def __init__(self, *args, **kwargs):
```
Comment (GittiHab) [inline Tests/MetaPathRatingGraphTest.py]: Please use the `self.setUp` method instead.

File: Tests/MetaPathRatingGraphTest.py [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+        super(MetaPathRatingTest, self).__init__(*args, **kwargs)
+        self.meta_paths = self.create_meta_path_list()
+
+    def test_creation(self):
+        rating_graph = MetaPathRatingGraph()
+
+        self.assertTrue(rating_graph is not None)
```
Comment (GittiHab) [inline Tests/MetaPathRatingGraphTest.py]: There is a method `assertNotNone`, as far as I know.

File: Tests/MetaPathRatingGraphTest.py [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+        expected_distances = {
+            self.meta_paths[1]: [0.0, 4.2, 10.2, 11.2, 3.0],
+            self.meta_paths[2]: [0.0, 6.0, 7.0],
+            self.meta_paths[3]: [0.0, 1],
+            self.meta_paths[5]: [0.0],
+            self.meta_paths[4]: [0.0]
+        }
+
+        for node in rating_graph.all_nodes():
+            rating_graph_result = rating_graph.all_pair_distances()[rating_graph.meta_paths_map[node]]
```
Comment (GittiHab) [inline Tests/MetaPathRatingGraphTest.py]: Do we need to compute all the distances in **every** iteration?

File: util/datastructures.py [hunk 15]
```diff
@@ -56,3 +57,68 @@ def set_distances(self, distances: List[float]) -> None:
             distances), 'Number of meta-paths which is {} doesn\'t match number of passed distances which is {}'.format(
             len(self.meta_paths), len(distances))
         self.distances = distances
+
+
+class MetaPathRatingGraph:
+
+    def __init__(self):
+        self.g = Graph(directed=True)
```
Comment (GittiHab) [inline util/datastructures.py]: Can we still rename this to graph?

File: util/datastructures.py [hunk 16]
```diff
@@ -56,3 +57,68 @@ def set_distances(self, distances: List[float]) -> None:
...
+        self.distance = self.g.new_edge_property("double")
+        self.meta_path_to_vertex = {}
+        self.vertex_to_meta_path = {}
+
+
+    def __add_meta_path(self, meta_path: MetaPath):
```
Comment (GittiHab) [inline util/datastructures.py]: Please always specify the return types

File: util/datastructures.py [hunk 17]
```diff
@@ -56,3 +57,68 @@ def set_distances(self, distances: List[float]) -> None:
...
+        """
+        :param a: The index of the meta-path will be retrieved or the meta-path is mapped to a new index.
```
Comment (GittiHab) [inline util/datastructures.py]: There is no description of the return value

File: util/datastructures.py [hunk 18]
```diff
@@ -56,3 +57,68 @@ def set_distances(self, distances: List[float]) -> None:
...
+            meta_path_vertex = self.meta_path_to_vertex[meta_path]
+        else:
+            meta_path_vertex = self.g.add_vertex()
+            self.meta_path_to_vertex[meta_path] = meta_path_vertex
+            self.vertex_to_meta_path[meta_path_vertex] = meta_path
+        return meta_path_vertex
+
+
+
+    def add_user_rating(self, superior_meta_path: MetaPath, inferior_meta_path: MetaPath, distance: float):
```
Comment (GittiHab) [inline util/datastructures.py]: Return type missing...

File: util/datastructures.py [hunk 19]
```diff
@@ -56,3 +57,68 @@ def set_distances(self, distances: List[float]) -> None:
...
+        """
+           :param a: The meta-path, which was rated higher compared to b.
+           :param b: The meta-path, which was rated lower compared to a.
+           :param distance: The distance between meta-paths a and b.
```
Comment (GittiHab) [inline util/datastructures.py]: Return value description missing...

**Comment by GittiHab** [comment]
Graph-tool is missing as a travis requirement.

**Comment by MkuuWaUjinga** [comment]
All comments are integrated now. 
Only big problem: Have failed to install graph_tool on travis. That's the reason why the builds error.

**Comment by Baschdl** [comment]
Travis just updated their Ubuntu 14.04 container as the newest container and don’t support newer Ubuntu versions. Graph-tool needs 16.04 to run. So probably we’ve to switch to Docker container testing on travis.

**Comment by Baschdl** [comment]
#12 should fix all your problems

**Pull request merged**